docs/src/docs/asciidoc/end-to-end-gradle-guide.adoc

```diff
@@ -0,0 +1,871 @@
+
+In some cases agent may include more metadata than it is actually needed. You can filter metadata using the agent filter files.
+These filter files that agent consumes have the following structure:
+
+[source,json,subs="verbatim,attributes", role="multi-language-sample"]
+----
+{
+ "rules": [
+    {"includeClasses": "some.class.to.include.**"},
+    {"excludeClasses": "some.class.to.exclude.**"},
+  ],
+  "regexRules": [
+    {"includeClasses": "regex\.example\.class.*"},
+    {"excludeClasses": "regex\.example\.exclude[0-9]+"},
+  ]
+}
+----
+
+The process how you can pass the config files to the agent is described in the <<additional-agent-options,previous section>>.
+
+We can see on the example how different filter files affect generated metadata:
+
+We are starting with the simple filter file:
+
+[source,json,subs="verbatim,attributes", role="multi-language-sample"]
+----
+{
+  "rules": [
+    {"includeClasses": "**"}
+  ]
+}
+----
+
+This filter file will instruct the agent to include everything and therefore, you will get a massive config file.
+For example this is how `reachability-metadata.json` looks like:
+
+[source,json,subs="verbatim,attributes", role="multi-language-sample"]
+----
+{
+  {
+    "reflection": [
+      {
+        "condition": {
+          "typeReached": "java.io.ObjectInputStream"
+        },
+        "type": "[Ljava.lang.Object;"
+      },
+      {
+        "condition": {
+          "typeReached": "java.io.ObjectInputStream"
+        },
+        "type": "java.util.LinkedHashSet"
+      },
+      {
+        "condition": {
+          "typeReached": "org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor"
+        },
+        "type": "org.example.NativeTests"
+      },
+      {
+        "condition": {
+          "typeReached": "org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor"
+        },
+        "type": "org.example.NativeTests",
+        "allDeclaredFields": true
+      },
+      {
+        "condition": {
+          "typeReached": "org.junit.jupiter.engine.descriptor.ExtensionUtils"
+        },
+        "type": "org.example.NativeTests"
+      },
+      ...
+    ],
+    "resources": [
+      {
+        "condition": {
+          "typeReached": "org.junit.platform.launcher.core.LauncherFactory"
+        },
+        "glob": "META-INF/services/org.junit.platform.engine.TestEngine"
+      },
+      {
+        "condition": {
+          "typeReached": "java.lang.ClassLoader"
+        },
+        "glob": "TestResource.txt"
+      },
+      ...
+    ],
+    "bundles": [],
+    "jni": [
+      {
+        "condition": {
+          "typeReached": "java.net.InetAddress"
+        },
+        "type": "java.lang.Boolean",
+        "methods": [
+          {
+            "name": "getBoolean",
+            "parameterTypes": [
+              "java.lang.String"
+            ]
+          }
+        ]
+      }
+    ]
+  }
+----
+
+As you can see, there are lots of entries that you may don't want.
```

Comment:
```suggestion
As you can see, there are lots of entries that you likely don't want.
They are present because the metadata was generated using tests, for example, and that the testing library was present at runtime during the agent execution pass, but in practice, your application will not use the testing library in production.
```